### ISSUE TITLE
Cannot attach undefined variable: T (different from #117 I think)

### ISSUE THREAD

**Issue by ljnelson** [raphw/byte-buddy#932]
(I'm aware of #117.)

I am running into this issue when working with a `MethodDescription` that represents a method whose return type is a parameterized type that has a type variable.  So, for example:

    public <T> Glop<T> getGlop() {
      return glop;
    }

I am getting this `MethodDescription` via an invocation that looks kind of like this:

    final MethodDescription methodDescription =
        aTypeDescriptionGeneric.getDeclaredMethods()
        .filter(is(new MethodDescription.Latent(aTypeDescriptionGeneric.asErasure(),
                                                aMethodDescriptionToken)));

I also get it with this:

    final MethodDescription methodDescription =
        aTypeDescriptionGeneric.getDeclaredMethods()
        .filter(hasSignature(aMethodDescriptionToken.asSignatureToken(aTypeDescriptionGeneric.asErasure())));

That is, someone has given me a `MethodDescription.Token` that (partially) describes the method in question; I'm trying to find the "real" method on the type.  When I debug, I notice that the `methodDescription` is indeed a `ForLoadedMethod` kind, which makes me think this is all working properly.  But then I get this error.  Isn't `getGlop()` a valid method?

The (relevant part of the) stack is here:
```
java.lang.IllegalArgumentException: Cannot attach undefined variable: T
		at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor$ForAttachment.onTypeVariable(TypeDescription.java:2003)
		at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor$ForAttachment.onTypeVariable(TypeDescription.java:1913)
		at net.bytebuddy.description.type.TypeDescription$Generic$OfTypeVariable$Symbolic.accept(TypeDescription.java:5938)
		at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor.onParameterizedType(TypeDescription.java:1850)
		at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor$ForAttachment.onParameterizedType(TypeDescription.java:1913)
		at net.bytebuddy.description.type.TypeDescription$Generic$OfParameterizedType.accept(TypeDescription.java:5149)
		at net.bytebuddy.description.method.MethodDescription$Latent.getReturnType(MethodDescription.java:1400)
```

That `MethodDescription$Latent` reference intrigues me since I've been working to eliminate `Latent` from my code (note my second example above which does not mention it).

Is this a bug, or is this error message trying to tell me something more important?

**Comment by ljnelson** [comment]
Debugging; I'll update this comment as I discover things.  Apologies if this causes notifications; I'll try to keep edits to a minimum.

https://github.com/raphw/byte-buddy/blob/3fef9b8ac7f44b85b1fc579eae60e843c0873d09/byte-buddy-dep/src/main/java/net/bytebuddy/description/type/TypeDescription.java#L1848-L1851

This seems to be where the problem is.  What I've found is that the `Latent` reference is not (thank goodness) my code but ByteBuddy code (I've had problems misusing `Latent` before).  So preliminary investigations seem to show this is a ByteBuddy bug.

If it helps, I think I'm triggering this bug by creating a `DynamicType` via a `DynamicType.Builder` that is set up to invoke an arbitrary method on another class.  That method is supplied as a token (as the issue description indicates).  I then use ByteBuddy to find the "real" method corresponding to the token, which it does.  The method in question, as noted in the issue description, definitely has a "free" type variable that it declares, no question.

During some hashcode calculations, ByteBuddy discovers that this method has a "free" type variable, and starts trying to find a `TypeVariableSource` to resolve it.  So it starts with the method itself, and of course discovers that the method's declaring class does not resolve this variable, and then it tries to see if the declaring class is enclosed, and it's not:

https://github.com/raphw/byte-buddy/blob/3fef9b8ac7f44b85b1fc579eae60e843c0873d09/byte-buddy-dep/src/main/java/net/bytebuddy/description/type/TypeDescription.java#L2000-L2007

I do not know the ByteBuddy internals at all.  But something seems a little wrong here.

**Comment by ljnelson** [comment]
I'm going to close this because I think I'm probably not using erasures properly somewhere and I don't want to bother you further until I have a reproducer or can confirm that the issue is my problem.

**Comment by ljnelson** [comment]
I will leave this closed: I'm reasonably sure this is because I am determining what sort of subclass to make given a `TypeDescription.Generic` that is handed to me—and so if the type contains type variables, I have to scan through it properly so that I can call `typeVariable()` on my `DynamicType.Builder`, which I was not doing.  This is of course [documented](https://javadoc.io/static/net.bytebuddy/byte-buddy/1.10.16/net/bytebuddy/ByteBuddy.html#subclass-net.bytebuddy.description.type.TypeDefinition-net.bytebuddy.dynamic.scaffold.subclass.ConstructorStrategy-); I just didn't notice.  (It does seem like an area where perhaps ByteBuddy could do this boring work for me, however.)

**Comment by raphw** [comment]
Sorry for not getting back earlier, I've been rather busy lately.

Byte Buddy needs to represent type variables as latent values internally due to the possibility of recurisve definitions (`T extends List<T>`). When you resolve a value, Byte Buddy needs to resolve the variable which requires processing the hierarchy. A type variable might be declared by the current method, it's declaring class and might refer subtypes etc. For computing the hash code, the erasure needs to be known etc what requires "attaching" the variable by locating the source.

The exception you see implies that a variable references a variable symbol that is not declared by anybody. You could create the following type as a class file:

````java
class Foo<T> {
  void bar(S s) { }
}
```

It's erased to a valid Java class but Byte Buddy (or the reflection API) could not make sense of it. Does the same error occur when reflecting on the code in question? If yes: you deal with an illegal class. If no: you found a bug and I'd love a reproducer.

**Comment by ljnelson** [comment]
First, you never have to apologize; this is an open-source project you make available to the rest of us out of your own benevolence.  It would be wrong of anyone to expect anything unless they're paying you!

Regarding this issue: it is indeed an issue with attachment.  I had (unintentionally) defined a subclass with "free" type variables, i.e. I subclassed a class that is parameterized and didn't "fill" all the "slots".  I resolved the problem by walking the hierarchy myself and made sure that my subclass declared its own type variables for all type variables that were not "filled".

The code that I wrote to do this was very boilerplate and generic which is why I was thinking that if a user calls `subclass(someParameterizedSuperType)` ByteBuddy could conveniently perform this boilerplate stuff and automatically add type variables as needed.
